gprofiler/utils/node.py

```diff
@@ -0,0 +1,223 @@
+
+    if (
+        "result" not in message.keys()
+        or "result" not in message["result"].keys()
+        or "type" not in message["result"]["result"].keys()
+        or message["result"]["result"]["type"] != "boolean"
+    ):
+        raise NodeDebuggerUnexpectedResponse(message)
+
+
+def _copy_module_into_process_ns(process: psutil.Process, musl: bool, version: str):
+    proc_root = get_proc_root_path(process)
+    libc = "musl" if musl else "glibc"
+    dest_inside_container = os.path.join("tmp", "node_module", libc, version)
+    dest = os.path.join(proc_root, dest_inside_container)
+    if os.path.exists(dest):
+        return "/" + dest_inside_container
+    src = resource_path(os.path.join("node", "module", libc, version))
+    shutil.copytree(src, dest)
+    add_permission_dir(dest, stat.S_IROTH, stat.S_IXOTH | stat.S_IROTH)
+    return "/" + dest_inside_container
+
+
+def _create_symlink_for_map(process: psutil.Process, pid_inside_ns: int):
+    proc_root = get_proc_root_path(process)
+    map_file_name = f"perf-{process.pid}.map"
+    src = os.path.join(proc_root, "tmp", f"perf-{pid_inside_ns}.map")
+    dest = os.path.join("/tmp", map_file_name)
+    os.symlink(src, dest)
+
+
+@lru_cache(1024)
+def _get_pid_inside_ns(pid: int) -> int:
+    with open(f"/proc/{pid}/status") as f:
+        for line in f.readlines():
+            if line.startswith("NSpid"):
+                return line.split("\t")[-1].strip("\n")
+
+
+def _generate_perf_map_wrapper(module_path: str) -> Callable:
+    def generate_perf_map():
+        debugger_url = _get_debugger_url()
+        sock = create_connection(debugger_url)
+        _load_dso(sock, module_path)
+
+    return generate_perf_map
+
+
+def _clean_up_wrapper(module_path: str, pid: int) -> Callable:
+    def clean_up():
+        debugger_url = _get_debugger_url()
+        sock = create_connection(debugger_url)
+        _stop_dso(sock, module_path)
+        shutil.rmtree(os.path.join("/tmp", "node_module"))
+        os.remove(os.path.join("/tmp", f"perf-{pid}.map"))
+
+    return clean_up
+
+
+def generate_map_for_node_processes():
+    """Iterates over all NodeJS processes, starts debugger for it, finds debugger URL,
+    copies node-linux-perf module into process' namespace, loads module and starts it.
+    After that it links it into gProfiler namespace's /tmp. This lets perf load it"""
+    processes = get_node_processes()
+    for process in processes:
+        try:
+            musl = is_musl(process)
+            node_version = get_exe_version(process)
+            node_major_version = node_version[1:].split(".")[0]
+            dest = _copy_module_into_process_ns(process, musl, node_major_version)
+            _start_debugger(process.pid)
+            pid_inside_ns = _get_pid_inside_ns(process.pid)
+            cp = run_in_ns(
+                ["pid", "mnt", "net"], _generate_perf_map_wrapper(dest), process.pid
+            )
+            if hasattr(cp, "stdout"):
```

Comment:
Are these relevant? Why would it have the `stdout` attribute?

Reply:
It looks like both `cp.stdout` and `cp.stderr` are empty, so we can remove it.